build.sbt

```diff
@@ -243,3 +249,134 @@ lazy val renaissance: Project = {
+  logger: Logger,
+  nonGpl: Boolean,
+  groupJars: Seq[(String, Seq[File], Seq[File])]
+) = {
+  val perProjectBenchmarkClasses = for ((project, allJars, loadedJars) <- groupJars) yield {
+    // Scan project jars for benchmarks and fill the property file.
+    logger.info(s"Generating JMH wrappers for project $project")
+    for {
+      info <- listBenchmarks(allJars, None)
+      if !nonGpl || info.distro() == License.MIT
+    } yield {
+      generateJmhWrapperBenchmarkClass(
+        info.benchClass.getSimpleName,
+        info.benchClass.getPackage.getName,
+        outputDir
+      )
+    }
+  }
+
+  perProjectBenchmarkClasses.flatten
+}
+
+// This project generates a custom JMH wrapper for each Renaissance benchmark.
+// Then, the project inserts JMH-specific functionality into the final JAR,
+// which can then be run in a JMH-compliant way.
+//
+// Here are several technical details.
+// First, to be able to generate the JMH wrapper classes, a project needs to "see"
+// all the projects that define the benchmarks.
+// Therefore, it was convenient to make the JMH project depend on the `renaissance` project.
+// Second, the sbt-jmh plugins adds some (unnecessary) extra functionality,
+// which forces this project to depend on the Scala library.
+// This is why we cannot set `autoScalaLibrary` to `false` here.
+// Third, the sbt-assembly plugin consequently automatically includes the Scala library
+// into the fat JAR, which breaks classloading in the benchmarks that have
+// a different Scala version (because the fat JAR is in the `AppClassLoader`).
+// To amend this, we forcefully remove the Scala library classfiles from the fat JAR
+// with a custom merge strategy -- this is fine, because the Scala library is only needed
+// by the extra functionality that the sbt-jmh inserts into the artifact, and we don't
+// need that anyway.
+lazy val renaissanceJmh = (project in file("renaissance-jmh"))
+  .enablePlugins(JmhPlugin)
+  .settings(
+    name := "renaissance-jmh",
+    version := (version in renaissance).value,
+    organization := (organization in renaissance).value,
+    scalafmtConfig := Some(file(".scalafmt.conf")),
+    nonGplOnly := (nonGplOnly in renaissance).value,
+    mainClass in assembly := Some("org.openjdk.jmh.Main"),
+    sourceGenerators in Compile := Def.taskDyn {
+      val log = streams.value.log
+      val outputDir = sourceManaged.in(Compile).value
+      val nonGpl = nonGplOnly.value
+      projectJars.map { groupJars =>
+        generateJmhWrapperBenchmarkClasses(outputDir, log, nonGpl, groupJars)
+      }
+    }.taskValue +: (sourceGenerators in Compile).value,
+    assembly in Jmh := ((assembly in Jmh) dependsOn (compile in Jmh)).value,
+    assemblyMergeStrategy in assembly := {
+      case PathList("scala", xs @ _*) =>
+        MergeStrategy.discard
+      case x =>
+        val oldStrategy = (assemblyMergeStrategy in assembly).value
+        oldStrategy(x)
+    }
+  )
+  .dependsOn(
+    renaissance
+  )
```

Comment:
Do you think that these chunks of JMH-specific code live in a separate `.scala` file (say `jmh.scala`) in the project directory?

Reply:
That's a good idea - I will move them.

Reply:
Moved JMH-specific functionality to `jmh.scala`.